proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,466 @@
+matrices. This is a useful feature that users would like to use.
+
+## High-level description
+
+The solution to this problem is to add a new class,
+`vk::khr::CooperativeMatrix`, which will be defined in a header file
+"vk/khr/cooperative_matrix.h". This class will create an object with SPIR-V type
+`OpTypeCooperativeMatrixKHR`. Functions are added that will expose the SPIR-V
+operations that take cooperative matrices as operands. All functions are defined
+to match the corresponding operations in the
+(SPV_KHR_cooperative_matrix)[https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_cooperative_matrix.html]
+extension.
+
+To help implement `vk::khr::CooperativeMatrix`, we will also introduce utility
+classes `vk::util::ArithmeticSelector` and `vk::util::ConversionSelector`. This
+class will be defined in a header file "vk/opcode_selector.h". They can be used
+to generate inline SPIR-V arithmetic and conversion instructions with the
+correct opcode for the types.
+
+The interface for `vk::khr::CooperativeMatrix` uses enums that are defined in
+the SPIR-V specification. The required enums will be defined in "vk/spirv.h".
+
+## Detailed design
+
+### `vk::khr::CooperativeMatrix`
+
+The `CooperativeMatrix` class will be defined in a header file as a wrapper
+around a `vk::SpirvType` that will expand to the appropriate SPIR-V type. This
+class will have the following interface.
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+          CooperativeMatrixUse use>
+class CooperativeMatrix {
+  template <class NewComponentType>
+  CooperativeMatrix<NewComponentType, scope, rows, columns, use> cast();
+
+  // Apply OpSNegate or OFNegate, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix negate();
+
+  // Apply OpIAdd or OFAdd, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator+(CooperativeMatrix other);
+
+  // Apply OpISub or OFSub, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator-(CooperativeMatrix other);
+
+  // Apply OpIMul or OFMul, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator*(CooperativeMatrix other);
+
+  // Apply OpSDiv, OpUDiv or OFDiv, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix operator/(CooperativeMatrix other);
+
+  // Apply OpMatrixTimesScalar in a element by element manner.
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the workgroup scope
```

Comment:
The Av/Vis bits are backward - stores should do MakePointerAvailable, loads should do MakePointerVisible.

Reply:
done.